### PR TITLE
[Torch Dialect] Decompose AtenTriuOp

### PR DESCRIPTION
**Pull request by Mi-Jiazhi**
decompose like:
```
import torch

def my_triu(x, diag):
    rows = torch.ops.aten.size(x, -2)
    cols = torch.ops.aten.size(x, -1)

    row_indices = torch.ops.aten.arange(rows).unsqueeze(1)
    col_indices = torch.ops.aten.arange(cols).unsqueeze(0)

    cond = torch.ops.aten.ge(
        col_indices, torch.ops.aten.add(row_indices, diag))
    return torch.ops.aten.where(cond, x, 0)

x = torch.rand(5, 7)
assert torch.allclose(my_triu(x, 0), torch.triu(x, 0))
assert torch.allclose(my_triu(x, 1), torch.triu(x, 1))
assert torch.allclose(my_triu(x, 2), torch.triu(x, 2))
assert torch.allclose(my_triu(x, -1), torch.triu(x, -1))
```

### CODE REVIEW

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/elementwise_comparison.py [hunk 1]
```diff
@@ -795,3 +795,26 @@ def forward(self, x):
...
+    @annotate_args([
+        None,
+        ([4,5], torch.float32, True),
+    ])
+    def forward(self, x):
+        return torch.ops.aten.triu(x, 1)
+
+
+@register_test_case(module_factory=lambda: TriuModule())
+def TriuModule_basic(module, tu: TestUtils):
```
Comment (qingyunqu) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/elementwise_comparison.py:815]: `TriuModule` => `AtenTriuModule`

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 2]
```diff
@@ -246,6 +246,56 @@ class DecomposeAtenAmaxOp : public OpRewritePattern<AtenAmaxOp> {
...
+      op->emitError("non-const diagonal parameter unsupported");
+      return failure();
+    }
+
+    auto inputSize = inputType.getSizes();
+    auto rowNum = *(inputSize.rbegin() + 1);
+    auto colNum = *inputSize.rbegin();
+
+    SmallVector<bool> condVal(rowNum * colNum, 0);
+    // bool *condVal = new bool[rowNum * colNum];
```
Comment (qingyunqu) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:275]: remove this comment

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 3]
```diff
@@ -246,6 +246,56 @@ class DecomposeAtenAmaxOp : public OpRewritePattern<AtenAmaxOp> {
...
+    int64_t diagonal;
+    if (!m_TorchConstantInt(&diagonal).match(
+            op.getDiagonal().getDefiningOp())) {
+      op->emitError("non-const diagonal parameter unsupported");
+      return failure();
+    }
+
+    auto inputSize = inputType.getSizes();
+    auto rowNum = *(inputSize.rbegin() + 1);
+    auto colNum = *inputSize.rbegin();
```
Comment (qingyunqu) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:272]: only support tensor with 2 dimensions?  You should check this like constant diagonal.

**Review by qingyunqu — CHANGES_REQUESTED** [review]

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 4]
```diff
@@ -246,6 +246,55 @@ class DecomposeAtenAmaxOp : public OpRewritePattern<AtenAmaxOp> {
...
+    Value input = op.getSelf();
+    auto inputType = input.getType().dyn_cast<BaseTensorType>();
+    if (!inputType.areAllSizesKnown() || !inputType.hasDtype()) {
+      op->emitError("should have static shape and dtype");
+      return failure();
+    }
+
+    int64_t diagonal;
+    if (!m_TorchConstantInt(&diagonal).match(
+            op.getDiagonal().getDefiningOp())) {
```
Comment (qingyunqu) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:265]: => `if (!matchPattern(op.getDiagonal(), m_TorchConstantInt(&diagonal)) {`

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 5]
```diff
@@ -246,6 +246,55 @@ class DecomposeAtenAmaxOp : public OpRewritePattern<AtenAmaxOp> {
...
+      op->emitError("non-const diagonal parameter unsupported");
+      return failure();
+    }
+
+    auto inputSize = inputType.getSizes();
```
Comment (qingyunqu) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:270]: I think you should check the `inputType` should have rank >= 2.

**Review by qingyunqu — CHANGES_REQUESTED** [review]

**Review by qingyunqu — APPROVED** [review]
LGTM

**Comment by qingyunqu** [comment]
@ramiro050 Could you please review this PR?

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 6]
```diff
@@ -246,6 +246,56 @@ class DecomposeAtenAmaxOp : public OpRewritePattern<AtenAmaxOp> {
...
+class DecomposeAtenTriuOp : public OpRewritePattern<AtenTriuOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AtenTriuOp op,
+                                PatternRewriter &rewriter) const override {
+    Location loc = op.getLoc();
+    Value input = op.getSelf();
+    auto inputType = input.getType().cast<BaseTensorType>();
+    if (!inputType.areAllSizesKnown() || !inputType.hasDtype()) {
+      return rewriter.notifyMatchFailure(op, "should have static shape and dtype");
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:259]: I think we can have dynamic support for this op. Here is a potential decomposition:

```python
import torch

def my_triu(x, diag):
    rows = torch.ops.aten.size(x, -2)
    cols = torch.ops.aten.size(x, -1)

    row_indices = torch.ops.aten.arange(rows).unsqueeze(1)
    col_indices = torch.ops.aten.arange(cols).unsqueeze(0)

    cond = torch.ops.aten.ge(
        col_indices, torch.ops.aten.add(row_indices, diag))
    return torch.ops.aten.where(cond, x, 0)

x = torch.rand(5, 7)
assert torch.allclose(my_triu(x, 0), torch.triu(x, 0))
assert torch.allclose(my_triu(x, 1), torch.triu(x, 1))
assert torch.allclose(my_triu(x, 2), torch.triu(x, 2))
assert torch.allclose(my_triu(x, -1), torch.triu(x, -1))

**Review by ramiro050 — CHANGES_REQUESTED** [review]

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/elementwise.py [hunk 7]
```diff
@@ -3251,6 +3251,52 @@ def AtenTriuWithPosDiagonalModule_basic(module, tu: TestUtils):
...
+
+
+class TriuBroadcastModule(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+
+    @export
+    @annotate_args([
+        None,
+        ([3,4,5,6], torch.float32, True),
```
Comment (ramiro050) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/elementwise.py:3286]: Can we also have a test with dynamic shapes?

**Review by ramiro050 — APPROVED** [review]
LGTM, just one request

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/elementwise.py [hunk 8]
```diff
@@ -3251,6 +3251,52 @@ def AtenTriuWithPosDiagonalModule_basic(module, tu: TestUtils):
...
+
+@register_test_case(module_factory=lambda: TriuBroadcastModule())
+def TriuBroadcastModule_basic(module, tu: TestUtils):
+    module.forward(tu.rand(3,4,5,6))
+
+
+# ==============================================================================
+
+
 class AtenTriuWithNegDiagonalModule(torch.nn.Module):
```
Comment (qingyunqu) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/elementwise.py:3300]: It already has a dynamic shape case here, cc @ramiro050 .

**Pull request merged**